core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -335,6 +336,44 @@ private[spark] object Utils extends Logging {
     }
   }
 
+  /**
+   * Copy all data from an InputStream to an OutputStream upto maxSize and
+   * close the input stream if all data is read.
+   * @return A tuple of boolean, which is whether the stream was fully copied, and an InputStream,
+   *         which is a combined stream of read data and any remaining data
+   */
+  def copyStreamUpTo(in: InputStream, maxSize: Long): (Boolean, InputStream) = {
+    var count = 0L
+    val out = new ChunkedByteBufferOutputStream(64 * 1024, ByteBuffer.allocate)
+    val streamCopied = tryWithSafeFinally {
```

Comment:
nit: I think a better name for `streamCopied` would be `fullyCopied` ... everytime I come back to this code I always get a bit confused that maybe no copying is happening in some situations.